### ISSUE TITLE
Document undocumented items

### ISSUE THREAD

**Issue by mankinskin** [seanmonstar/warp#742]
Many code items are not documented by the documentation, which makes the API harder to understand. Examples are 
- [`Func`](https://docs.rs/warp/0.2.5/src/warp/generic.rs.html#38)
- ['CombinedRejection`](https://docs.rs/warp/0.2.5/src/warp/reject.rs.html#592)
- [`FilterBase`](https://docs.rs/warp/0.2.5/src/warp/filter/mod.rs.html#38)

I don't know the rationale behind hiding these items, but it would make the crate a lot easier to work with if it was fully documented.

**Comment by jxs** [comment]
Hi, those traits are public only because `Filter` leaks them, but they are not meant to be used as building blocks for warp applications, `FilterBase` kinda refers that, and [here](https://docs.rs/warp/0.2.5/src/warp/filter/mod.rs.html#163) you have more info about it, but yeah maybe we could explicitly state that they are internal to warp

**Comment by mankinskin** [comment]
But they are not internal to warp, because Filter leaks them and they are mentioned in the documentation. You need to know about these items to use Filter properly.

**Comment by guswynn** [comment]
I agree with this, I am trying to figure out how Warp actually works. @jxs could we expose FilterBase as `pub trait FilterBase: Sealed` and, to prevent semver issues, seal it with a canonical pub-trait+private-path `Sealed` trait. We can do the same thing with HList and friends, and then we can add normal doc comments to these traits

**Comment by Stargateur** [comment]
I also found this very annoying, this try to bend the rule of pub and private of Rust. Doesn't help the user and just make life complicated to try to avoid breaking change.

`FilterBase` contains associated type that the user NEED to know to use `Filter`.
